core/src/main/java/org/elasticsearch/monitor/fs/FsService.java

```diff
@@ -19,54 +19,71 @@
 
-import org.elasticsearch.common.component.AbstractComponent;
+import org.elasticsearch.common.component.AbstractLifecycleComponent;
 import org.elasticsearch.common.settings.Setting;
 import org.elasticsearch.common.settings.Setting.Property;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.unit.TimeValue;
-import org.elasticsearch.common.util.SingleObjectCache;
+import org.elasticsearch.common.util.concurrent.FutureUtils;
 import org.elasticsearch.env.NodeEnvironment;
 
 import java.io.IOException;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiFunction;
 
-/**
- */
-public class FsService extends AbstractComponent {
+public class FsService extends AbstractLifecycleComponent<FsService> {
 
     private final FsProbe probe;
-
-    private final SingleObjectCache<FsInfo> fsStatsCache;
+    private final TimeValue refreshInterval;
+    private final BiFunction<Runnable, TimeValue, ScheduledFuture> scheduler;
+    private final AtomicReference<FsInfo> ref = new AtomicReference<>();
+    private final Runnable probeRunner;
+    private ScheduledFuture scheduledFuture;
 
     public final static Setting<TimeValue> REFRESH_INTERVAL_SETTING =
         Setting.timeSetting("monitor.fs.refresh_interval", TimeValue.timeValueSeconds(1), TimeValue.timeValueSeconds(1),
             Property.NodeScope);
 
-    public FsService(Settings settings, NodeEnvironment nodeEnvironment) throws IOException {
+    public FsService(
+            final Settings settings,
+            final NodeEnvironment nodeEnvironment,
+            final BiFunction<Runnable, TimeValue, ScheduledFuture> scheduler) throws IOException {
         super(settings);
         this.probe = new FsProbe(settings, nodeEnvironment);
-        TimeValue refreshInterval = REFRESH_INTERVAL_SETTING.get(settings);
-        fsStatsCache = new FsInfoCache(refreshInterval, probe.stats());
+        refreshInterval = REFRESH_INTERVAL_SETTING.get(settings);
+        this.scheduler = scheduler;
+        probeRunner = () -> {
+            try {
+                final FsInfo fsInfo = probe.stats(ref.get());
+                ref.set(fsInfo);
+            } catch (IOException e) {
+            }
+        };
         logger.debug("using refresh_interval [{}]", refreshInterval);
     }
 
     public FsInfo stats() {
-        return fsStatsCache.getOrRefresh();
+        return ref.get();
     }
 
-    private class FsInfoCache extends SingleObjectCache<FsInfo> {
-        public FsInfoCache(TimeValue interval, FsInfo initValue) {
-            super(interval, initValue);
-        }
+    @Override
+    protected void doStart() {
+        // force the probe to run in case stats is called before the
+        // first scheduled run occurs
+        probeRunner.run();
+        // now schedule the probe to run periodically
+        scheduledFuture = scheduler.apply(probeRunner, refreshInterval);
```

Comment:
Can we give some insight into why we are doing this thread scheduling? What makes this io stats special as opposed to other stats where its either raw numbers, or in the sigar days, e.g. based on the last time you called it.

As far as returning the relative measures, I think we should consider that too. I think its strange to suddenly provide a ton of _relative_ stats here when in general we dont do that (we provide the raw numbers instead). It puts the onus on us to be doing sampling/scheduling when in fact that might not even be wanted, and it introduces a whole shit ton of floating point (and nobody understands floating point) variables that I feel will be problematic, as far as then doing any kinds of graphs or anything on top of them.

I do think if its easy, we should provide "iops" (say computed based on the last time we called it) but its just to have one simple number for people to look at. Otherwise for the more detailed stuff we provide the "raw" counts?